File: algorithms/663. Equal Tree Partition/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def checkEqualTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def treesum(node):
            if not node:
                return 0
            s = node.val + treesum(node.left) + treesum(node.right)
            counter[s] += 1
            return s

        counter = collections.defaultdict(int)
        s = treesum(root)
        counter[s] -= 1
        return s % 2 == 0 and counter[s / 2] > 0

# 194 / 194 test cases passed.
# Status: Accepted
# Runtime: 152 ms

File: algorithms/663. Equal Tree Partition/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool checkEqualTree(TreeNode* root) {
        unordered_map<int, int> m;
        int s = treesum(root, m);
        --m[s];
        return s % 2 == 0 and m[s/2] > 0;
    }

    int treesum(TreeNode * node, unordered_map<int, int> & m) {
        if ( nullptr == node )
            return 0;
        int s = node->val + treesum(node->left, m) + treesum(node->right, m);
        ++m[s];
        return s;
    }
};

// 194 / 194 test cases passed.
// Status: Accepted
// Runtime: 39 ms
